placeholder-transaction-trait/src/message_trait.rs

```diff
@@ -0,0 +1,217 @@
+use {
+    crate::{Instruction, MessageAddressTableLookup},
+    core::fmt::Debug,
+    solana_sdk::{
+        feature_set::FeatureSet,
+        hash::Hash,
+        message::{AccountKeys, TransactionSignatureDetails},
+        nonce::NONCED_TX_MARKER_IX_INDEX,
+        precompiles::{get_precompiles, is_precompile},
+        pubkey::Pubkey,
+        sysvar::instructions::{BorrowedAccountMeta, BorrowedInstruction},
+        transaction::TransactionError,
+    },
+};
+
+mod reference;
+mod sanitized_message;
+mod sanitized_transaction;
+
+// - Clone to support possible cloning from holding in Vec
+// - Debug to support legacy logging
+pub trait MessageTrait: Clone + Debug {
+    /// Return the number of signatures in the message.
+    fn num_signatures(&self) -> u64;
+
+    /// Returns the number of requested write-locks in this message.
+    /// This does not consider if write-locks are demoted.
+    fn num_write_locks(&self) -> u64;
+
+    /// Return the recent blockhash.
+    fn recent_blockhash(&self) -> &Hash;
+
+    /// Return the number of instructions in the message.
+    fn num_instructions(&self) -> usize;
+
+    /// Return an iterator over the instructions in the message.
+    fn instructions_iter(&self) -> impl Iterator<Item = Instruction>;
+
+    /// Return an iterator over the instructions in the message, paired with
+    /// the pubkey of the program.
+    fn program_instructions_iter(&self) -> impl Iterator<Item = (&Pubkey, Instruction)>;
+
+    /// Return the account keys.
+    fn account_keys(&self) -> AccountKeys;
+
+    /// Return the fee-payer
+    fn fee_payer(&self) -> &Pubkey;
+
+    /// Returns `true` if the account at `index` is writable.
+    fn is_writable(&self, index: usize) -> bool;
+
+    /// Returns `true` if the account at `index` is signer.
+    fn is_signer(&self, index: usize) -> bool;
+
+    /// Returns true if the account at the specified index is invoked as a
+    /// program in top-level instructions of this message.
+    fn is_invoked(&self, key_index: usize) -> bool;
+
+    /// Returns true if the account at the specified index is an input to some
+    /// program instruction in this message.
+    fn is_instruction_account(&self, key_index: usize) -> bool {
+        if let Ok(key_index) = u8::try_from(key_index) {
+            self.instructions_iter()
+                .any(|ix| ix.accounts.contains(&key_index))
+        } else {
+            false
+        }
+    }
+
+    /// Return signature details.
+    fn get_signature_details(&self) -> TransactionSignatureDetails {
```

Comment:
Should this just be a free + generic function?

Leaving it on trait gives us flexibility to chagne the impl; don't anticipate doing that rn, but possibly could just store these details internally and copy it.

Reply:
removed this; can easily have a free-function that replaces the needed functionality, want to keep the interface here to a minimum.